File: concours/logx_provenance.js
// ─── Provenance par champ ────────────────────────────────────────────────────
// Pour la confiance : d'où vient chaque donnée enrichie de l'indicatif en cours
// de saisie. Un petit relevé (Pays · cty.dat, Distance · calculé…) sous la zone
// d'enrichissement. LECTURE SEULE, non-intrusif : écoute la frappe de l'indicatif
// (débounce), ne touche NI aux champs de saisie NI au chemin critique.

(function(global){
  'use strict';

  var _deb = null;

  function esc(s){
    return String(s == null ? '' : s)
      .replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;').replace(/'/g, '&#39;');
  }

  function _rendre(rows){
    var el = document.getElementById('provenancePanel');
    if(!el) return;
    rows = rows || [];
    if(!rows.length){ el.hidden = true; el.innerHTML = ''; return; }
    el.hidden = false;
    el.innerHTML = '<div class="prov-h">Provenance</div>' + rows.map(function(r){
      return '<div class="prov-row">' +
        '<span class="prov-c">' + esc(r.champ) + '</span>' +
        '<span class="prov-v">' + esc(r.valeur) + '</span>' +
        '<span class="prov-s">' + esc(r.source) + '</span>' +
        '</div>';
    }).join('');
  }

  function _maj(call){
    call = String(call || '').trim();
    if(call.length < 3){ _rendre([]); return; }   // trop court pour résoudre
    fetch('/calldb/provenance?call=' + encodeURIComponent(call))
      .then(function(r){ return r.ok ? r.json() : null; })
      .then(function(d){ _rendre(d && d.rows); })
      .catch(function(){});
  }

  // S'attache À SA PROPRE écoute de l'indicatif (n'altère pas les gestionnaires
  // existants). Débounce pour ne pas interroger à chaque touche.
  function _brancher(){
    var inp = document.getElementById('inputCall');
    if(!inp) return;
    inp.addEventListener('input', function(){
      if(_deb) clearTimeout(_deb);
      var v = inp.value;
      _deb = setTimeout(function(){ _maj(v); }, 500);
    });
  }

  global.LogxProvenance = { _rendre: _rendre, _maj: _maj, _brancher: _brancher };

  if(typeof document !== 'undefined' && typeof fetch === 'function'){
    if(document.readyState === 'loading') document.addEventListener('DOMContentLoaded', _brancher);
    else _brancher();
  }

})(typeof window !== 'undefined' ? window : this);

File: concours/logx_provenance.py
# -*- coding: utf-8 -*-
"""Provenance par champ : d'où vient chaque donnée enrichie d'un indicatif.

Pour la confiance (« FAIT »), on montre la SOURCE de chaque valeur. v1, à partir
de l'indicatif SEUL et hors-ligne :
  - Pays / Continent / Zone CQ / Zone ITU -> cty.dat (base DXCC locale) ;
  - Distance / Azimut -> calculés (position de référence cty.dat + votre QTH).

Fonction pure et déterministe (aucun réseau) ; le locator/nom via callbook
pourront s'ajouter plus tard (source = fournisseur callbook). Lecture seule.
"""


def provenance(call, cfg=None):
    """[{champ, valeur, source}] pour l'indicatif `call`. Vide si inconnu."""
    import logx_dxcc as dxcc
    from logx_utils import locator_to_latlon, haversine, bearing

    call = str(call or '').strip()
    rows = []
    if not call:
        return rows
    info = dxcc.lookup(call)
    if not info:
        return rows

    if info.get('country'):
        rows.append({'champ': 'Pays', 'valeur': info['country'], 'source': 'cty.dat'})
    if info.get('continent'):
        rows.append({'champ': 'Continent', 'valeur': str(info['continent']), 'source': 'cty.dat'})
    if info.get('cq_zone') is not None:
        rows.append({'champ': 'Zone CQ', 'valeur': str(info['cq_zone']), 'source': 'cty.dat'})
    if info.get('itu_zone') is not None:
        rows.append({'champ': 'Zone ITU', 'valeur': str(info['itu_zone']), 'source': 'cty.dat'})

    my_loc = str((cfg or {}).get('locator') or '').strip()
    if info.get('lat') is not None and info.get('lon') is not None and my_loc:
        mll = locator_to_latlon(my_loc)
        if mll and mll[0] is not None:
            d = haversine(mll[0], mll[1], info['lat'], info['lon'])
            az = bearing(mll[0], mll[1], info['lat'], info['lon'])
            rows.append({'champ': 'Distance', 'valeur': '%d km' % round(d), 'source': 'calculé (cty.dat)'})
            rows.append({'champ': 'Azimut', 'valeur': '%d°' % round(az), 'source': 'calculé (cty.dat)'})
    return rows
